Editorial review: Document element-scoped view transitions (#44217)
mdn/content: files/en-us/web/css/reference/properties/view-transition-scope/index.md

---
title: "`view-transition-scope` CSS property"
short-title: view-transition-scope
slug: Web/CSS/Reference/Properties/view-transition-scope
page-type: css-property
browser-compat: css.properties.view-transition-scope
sidebar: cssref
---

> The **`view-transition-scope`** [CSS](/en-US/docs/Web/CSS) property enables the discoverability of elements with {{cssxref("view-transition-name")}} values set on them (for the purposes of creating view transition [snapshots](/en-US/docs/Web/API/View_Transition_API/Using#an_aside_on_snapshots)) to be isolated to a specific element subtree.

Reviewer: it's a long sentence. can we split it and make it more compact like this:
```suggestion
The **`view-transition-scope`** [CSS](/en-US/docs/Web/CSS) property enables the discoverability of elements with {{cssxref("view-transition-name")}} values set on them. This limits the element subtree that is used for creating [snapshots](/en-US/docs/Web/API/View_Transition_API/Using#an_aside_on_snapshots) during a view transition.
```

Author: This doesn't quite make sense. I've kept the parens but cut it down a bit:

> The **`view-transition-scope`** [CSS](/en-US/docs/Web/CSS) property enables the discoverability of elements with {{cssxref("view-transition-name")}} values set on them (and therefore the creation of view transition [snapshots](/en-US/docs/Web/API/View_Transition_API/Using#an_aside_on_snapshots)) to be isolated to a specific element subtree.

## Syntax

```css
/* Keyword values */
view-transition-scope: none;
view-transition-scope: all;

/* Global values */
view-transition-scope: inherit;
view-transition-scope: initial;
view-transition-scope: revert;
view-transition-scope: revert-layer;
view-transition-scope: unset;
```

### Values

- `none`

>   - : The initial value. Discoverability of elements to snapshot during a view transition is not limited to a selected element's subtree.

Reviewer: ```suggestion
  - : The initial value. Discoverability of elements to snapshot during a view transition is not limited to any specific subtree.
```

Author: updated

- `all`

>   - : Discoverability of all snapshottable elements (with any non-`none` {{cssxref("view-transition-name")}}) during a view transition is limited to a selected element's subtree.

Reviewer: or flip it around and also extract the parenthetical info into its own separate sentence...something like this?

```suggestion
  - : Limits the discoverability of elements to snapshot during a view transition to the subtree of the element on which this property is set. Only elements with a non-`none` {{cssxref("view-transition-name")}}) are considered.
```

Author: updated

## Description

> When a view transition is triggered, as part of [the process](/en-US/docs/Web/API/View_Transition_API/Using#the_view_transition_process), the browser captures snapshots of elements that have a non-`none` {{cssxref("view-transition-name")}} set on them. These snapshots are then animated via CSS animations.

Reviewer: or more direct
```suggestion
During a [view transition process](/en-US/docs/Web/API/View_Transition_API/Using#the_view_transition_process), the browser captures snapshots of elements that have a non-`none` {{cssxref("view-transition-name")}} set on them. These snapshots are then animated via CSS animations.
```

Author: updated

> One issue that can arise is naming collisions between different view transitioned elements. Under normal circumstances, you can't have the same {{cssxref("view-transition-name")}} set on multiple elements — when the {{domxref("Element.startViewTransition()")}} function is called to start the transition, the browser throws an `InvalidStateError`.

Reviewer: "transitioned" -> elements undergoing/involved in a view transition?
"Under normal circumstances" -> by default? or normally?
```suggestion
One issue that can arise during this process is naming collisions between elements involved in a view transition. Normally, you can't have the same {{cssxref("view-transition-name")}} set on multiple elements — if you do, the browser throws an `InvalidStateError` when the {{domxref("Element.startViewTransition()")}} method is called to start the transition.
```

Author: updated

> You could solve this problem by setting a `view-transition-name` of `match-element` on the elements, to let the browser auto-assign internal unique names, but this won't work if you are including multiple components from different sources that you don't control. A naming collision might still occur.

Reviewer: ```suggestion
You could solve this problem by setting a `view-transition-name` of [`match-element`](/en-US/docs/Web/CSS/Reference/Properties/view-transition-name#match-element) on the elements to let the browser auto-assign internal unique names. However, this won't work if you are including multiple components from different sources that you don't control. A naming collision might still occur.
```

Author: updated

> The `view-transition-scope` property was implemented so that view transitions can be made self-contained. When `view-transition-scope: all` is set on an element, it limits the transition scope to that element and its descendants, which can be used to solve the above problem.

Reviewer: ```suggestion
The `view-transition-scope` property allows view transitions to be self-contained. When `view-transition-scope: all` is set on an element, it limits the transition scope to that element and its descendants, which can be used to solve the above problem.
```

Author: done

> Whenever a view transition is triggered, the browser automatically sets `view-transition-scope: all` on the transition scope's root element so that elements are only snapshotted and view transition animations applied inside the transition scope.

Reviewer: ```suggestion
Whenever a view transition is triggered, the browser automatically sets `view-transition-scope: all` on the transition root element, ensuring that only elements inside the transition scope are snapshotted and animated.
```

Author: done

## Formal definition

{{cssinfo}}

## Formal syntax

{{csssyntax}}

## Examples

> ### Demonstration of `view-transition-scope`

Reviewer: Using `view-transition-scope` to...

Author: updated

> This example demonstrates the effect of `view-transition-scope`.

Reviewer: expand on this a bit?

Author: done

#### HTML

> We include a {{htmlelement("button")}} element to control updating the DOM, plus several components with the class `change-me`, some of which are nested, all wrapped in a {{htmlelement("section")}} element.

Reviewer: ```suggestion
The HTML includes a {{htmlelement("button")}} element to control updating the DOM, plus several components with the class `change-me`, some of which are nested, all wrapped in a {{htmlelement("section")}} element.
```

Author: updated

```html live-sample___vt-scope
<button>Update DOM</button>
<section>
  <div class="change-me"><span>I can change</span></div>
  <div class="change-me">
    <span>I can change</span>
    <div class="change-me"><span>I can change</span></div>
  </div>
  <div class="change-me"><span>I can change</span></div>
</section>
```

#### CSS

> We start by setting the same `view-transition-name` on all components. We then set `view-transition-scope: all` on all of them, to isolate the view transition process for each one. We then set a longer {{cssxref("animation-duration")}} on all view transitions with this `view-transition-name` via the {{cssxref("::view-transition-group()")}} pseudo-element.

Reviewer: ```suggestion
We start by setting the same `view-transition-name` on all components. We then set `view-transition-scope: all` on all of them, to isolate the view transition process for each one. Next, we set a longer {{cssxref("animation-duration")}} on all view transitions with this `view-transition-name` via the {{cssxref("::view-transition-group()")}} pseudo-element.
```

Author: done

```css hidden live-sample___vt-scope
body {
  font: 1.2em / 1.5 sans-serif;
  width: 50%;
  max-width: 700px;
  margin: 0 auto;
}

section,
.change-me {
  border: 2px solid #666;
  padding: 10px;
}

section {
  background-color: orange;
}
```

```css live-sample___vt-scope
.change-me {
  background-color: white;
  view-transition-name: para-change;
  view-transition-scope: all;
}

::view-transition-group(para-change) {
  animation-duration: 1s;
}
```

#### JavaScript

> In our script, we start off by grabbing references to our button and the `<div>` elements (our components).

Reviewer: ```suggestion
The script starts by grabbing references to the button and the `<div>` elements (our components).
```

Author: done

```js live-sample___vt-scope
const btn = document.querySelector("button");
const divs = document.querySelectorAll("div");
```

Next, we define a function called `updateDivs()`, which toggles the text content of each component's nested {{htmlelement("span")}} element between two values, and also toggles the component's foreground and background colors between two values.

```js live-sample___vt-scope
function updateDivs() {
  divs.forEach((div) => {
    if (div.firstElementChild.textContent === "I can change") {
      div.firstElementChild.textContent = "I have changed";
      div.style.color = "white";
      div.style.backgroundColor = "black";
    } else {
      div.firstElementChild.textContent = "I can change";
      div.style.color = "black";
      div.style.backgroundColor = "white";
    }
  });
}
```

> Finally, we add a `click` event listener to the `<button>` element. When it is clicked, we first check whether `startViewtransition()` exists on the `document`

Reviewer: the line break is not needed here i guess
```suggestion
Finally, we add a `click` event listener to the `<button>` element. When the button is clicked, we first check whether `startViewTransition()` exists on the `document`
```

Author: done

> object — if not, we run `updateDivs()` and then `return` out of the function. This first part allows browsers that don't support view transitions to still update the DOM without error. Next, we run `updateDivs()` inside a `startViewTransition()` callback to trigger the view transition as the DOM updates.

Reviewer: ```suggestion
object — if it does not, we run `updateDivs()` and then `return` out of the function. This first part allows browsers that don't support view transitions to still update the DOM without error. Next, we run `updateDivs()` inside a `startViewTransition()` callback to trigger the view transition as the DOM updates.
```

Author: done

```js live-sample___vt-scope
btn.addEventListener("click", handleClick);

function handleClick(e) {
  if (!document.startViewTransition) {
    updateDivs();
    return;
  }
  document.startViewTransition(() => {
    updateDivs();
  });
}
```

#### Result

{{embedlivesample("vt-scope", "100%", 280)}}

> Click the button to see the view transition. Now try the following:

Reviewer: ```suggestion
Click the "Update DOM" button to see the view transition. Now try the following:
```

Author: updated

- Inspect one of the `<div>` elements.

> - In your browser DevTools' style panel, uncheck the `view-transition-scope: all;` declaration to unapply it.

Reviewer: ```suggestion
- In the Styles panel in your browser's developer tools, uncheck the `view-transition-scope: all;` declaration.
```

Author: updated

- Now switch to the JavaScript Console.

> - Press the "Update DOM" button again.

Reviewer: ```suggestion
- Click the "Update DOM" button again.
```

Author: done

> You should see that the view transition animation is not applied when the DOM changes, and you'll get an `InvalidStateError` in the console.

Reviewer: this should be indented to align with the previous bullet
```suggestion
You should see that the view transition animation is not applied when the DOM changes, and an `InvalidStateError` appears in the console.
```

also, the bullet points above should probably be numbered since they are steps to be followed in a sequence

Author: Updated! The bullet numbering is a good idea.

short-title: view-transition-scope
slug: Web/CSS/Reference/Properties/view-transition-scope
page-type: css-property
browser-compat: css.properties.view-transition-scope
sidebar: cssref
---

The **`view-transition-scope`** [CSS](/en-US/docs/Web/CSS) property enables the discoverability of elements with {{cssxref("view-transition-name")}} values set on them (for the purposes of creating view transition [snapshots](/en-US/docs/Web/API/View_Transition_API/Using#an_aside_on_snapshots)) to be isolated to a specific element subtree.

## Syntax

```css
/* Keyword values */
view-transition-scope: none;
view-transition-scope: all;

/* Global values */
view-transition-scope: inherit;
view-transition-scope: initial;
view-transition-scope: revert;
view-transition-scope: revert-layer;
view-transition-scope: unset;
```

### Values

- `none`
  - : The initial value. Discoverability of elements to snapshot during a view transition is not limited to a selected element's subtree.
- `all`
  - : Discoverability of all snapshottable elements (with any non-`none` {{cssxref("view-transition-name")}}) during a view transition is limited to a selected element's subtree.

## Description

> When a view transition is triggered, as part of [the process](/en-US/docs/Web/API/View_Transition_API/Using#the_view_transition_process), the browser captures snapshots of elements that have a non-`none` {{cssxref("view-transition-name")}} set on them. These snapshots are then animated via CSS animations. The `view-transition-scope` property allows you to limit the scope inside which the browser will search for elements to snapshot to a particular DOM tree.

Reviewer 2: The last sentence reads a bit awkward.   What does "limit the scope" really mean? I think we can be more concrete here without being too wordy.

The `view-transition-scope` property provides control over whether view-transition-names in an element's subtree are discoverable during the capture phase of a view-transition on an ancestor element or the document itself.

Author: The wording is a bit different now, but let me know if it still needs attention.

> When `view-transition-scope: all` is set on an element, it limits the scope of [element-scoped view transitions](/en-US/docs/Web/API/View_Transition_API/Using_element-scoped) to that element and its descendants. The default value, `view-transition-scope: none`, means that no scope limiting will take place.

Reviewer 2: view-transition-scope also applies to document view transitions.   Again the term "scope limited" is rather vague here.  When `view-transition-scope: all` is used, any view-transition-names in the element's subtree are not visible from the element's ancestors during the capture phase.

Author: Ah, got it. In my next commit, I've updated the page so that it isn't saying they are specific to element-scoped VTs.

> The `view-transition-scope` property was implemented so that element-scoped view transitions can be made self-contained. Whenever an element-scoped view transition is triggered, the browser automatically sets `view-transition-scope: all` on the transition scope's root element so that elements are only snapshotted and view transition animations applied inside the transition scope.

Reviewer 2: Setting `view-transition-scope: all` manually on a descendant element prior to starting a view-transition enables a running a view transition on the descendant concurrently without incurring a name collision.  

An example can be found in 
https://developer.chrome.com/docs/css-ui/view-transitions/element-scoped-view-transitions

Author: As mentioned over email, I have now added better content here. Let me know what you think.
